src/coreclr/src/jit/valuenum.cpp

```diff
@@ -8276,6 +8374,245 @@ void Compiler::fgValueNumberIntrinsic(GenTree* tree)
+        }
+
+        bool encodeResultType = vnEncodesResultTypeForSIMDIntrinsic(simdNode->gtSIMDIntrinsicID);
+
+        if (encodeResultType)
+        {
+            ValueNum vnSize     = vnStore->VNForIntCon(simdNode->gtSIMDSize);
+            ValueNum vnBaseType = vnStore->VNForIntCon(INT32(simdNode->gtSIMDBaseType));
+            ValueNum simdTypeVN = vnStore->VNForFunc(TYP_REF, VNF_SimdType, vnSize, vnBaseType);
+            resvnp.SetBoth(simdTypeVN);
+
+#ifdef DEBUG
+            if (verbose)
+            {
+                printf("    simdTypeVN is ");
+                vnPrint(simdTypeVN, 2);
+                printf("\n");
+            }
+#endif
+        }
+
+        VNFunc simdFunc = GetVNFuncForNode(tree);
+
+        if (tree->AsOp()->gtOp2 == nullptr)
+        {
+            // Unary SIMD nodes have a nullptr for op2.
+            excSetPair = op1Xvnp;
+            if (encodeResultType)
+            {
+                normalPair = vnStore->VNPairForFunc(tree->TypeGet(), simdFunc, op1vnp, resvnp);
+                assert(vnStore->VNFuncArity(simdFunc) == 2);
+            }
+            else
+            {
+                normalPair = vnStore->VNPairForFunc(tree->TypeGet(), simdFunc, op1vnp);
+                assert(vnStore->VNFuncArity(simdFunc) == 1);
+            }
+        }
+        else
+        {
+            ValueNumPair op2vnp;
+            ValueNumPair op2Xvnp;
+            vnStore->VNPUnpackExc(tree->AsOp()->gtOp2->gtVNPair, &op2vnp, &op2Xvnp);
+
+            excSetPair = vnStore->VNPExcSetUnion(op1Xvnp, op2Xvnp);
+            if (encodeResultType)
+            {
+                normalPair = vnStore->VNPairForFunc(tree->TypeGet(), simdFunc, op1vnp, op2vnp, resvnp);
+                assert(vnStore->VNFuncArity(simdFunc) == 3);
+            }
+            else
+            {
+                normalPair = vnStore->VNPairForFunc(tree->TypeGet(), simdFunc, op1vnp, op2vnp);
+                assert(vnStore->VNFuncArity(simdFunc) == 2);
+            }
+        }
+    }
+    tree->gtVNPair = vnStore->VNPWithExc(normalPair, excSetPair);
+}
+#endif // FEATURE_SIMD
+
+#ifdef FEATURE_HW_INTRINSICS
+// Does value-numbering for a GT_HWINTRINSIC node
+void Compiler::fgValueNumberHWIntrinsic(GenTree* tree)
+{
+    assert(tree->OperGet() == GT_HWINTRINSIC);
+    GenTreeHWIntrinsic* hwIntrinsicNode = tree->AsHWIntrinsic();
+    assert(hwIntrinsicNode != nullptr);
+
+    // For safety/correctness we must mutate the global heap valuenumber
+    //  for any HW intrinsic that performs a memory store operation
+    if (hwIntrinsicNode->OperIsMemoryStore())
+    {
+        fgMutateGcHeap(tree DEBUGARG("HWIntrinsic - MemoryStore"));
+    }
+
+    int      lookupNumArgs = HWIntrinsicInfo::lookupNumArgs(hwIntrinsicNode);
+    VNFunc   func          = GetVNFuncForNode(tree);
+    unsigned fixedArity    = vnStore->VNFuncArity(func);
+    //
+    // when the numArgs column in "hwinstrinsiclistxarch.h" is -1
+    // it gets translated into a value of 0
```

Comment:
I don't understand this comment. If it is `-1` that means it has a variable number of arguments. Where or in what way does it get translated into a value of 0?

Reply:
It getrs translated in thsi macro:

```
#define ValueNumFuncDef(vnf, arity, commute, knownNonNull, sharedStatic)                                               \
    if (commute)                                                                                                       \
        vnfOpAttribs[vnfNum] |= VNFOA_Commutative;                                                                     \
    if (knownNonNull)                                                                                                  \
        vnfOpAttribs[vnfNum] |= VNFOA_KnownNonNull;                                                                    \
    if (sharedStatic)                                                                                                  \
        vnfOpAttribs[vnfNum] |= VNFOA_SharedStatic;                                                                    \
    if (arity > 0)                                                                                                     \
        vnfOpAttribs[vnfNum] |= (arity << VNFOA_ArityShift);                                                           \
    vnfNum++;
```